File: #1048.longest-string-chain.cpp
class Solution {
public:
    int longestStrChain(vector<string>& words) {

        std::sort(words.begin(), words.end(), [](string& a, string& b){
            return a.size() < b.size();
        });

        unordered_map<string, int> longest;
        int result = 1;
        for (string& word : words) {
            for (int i = 0; i < word.size(); i++) {
                longest[word] = std::max(
                    longest[word],
                    longest[word.substr(0, i) + word.substr(i + 1)] + 1
                );
            }
            result = std::max(result, longest[word]);
        }
        return result;
    }
};

File: #1048.longest-string-chain.js
/**
 * @param {string[]} words
 * @return {number}
 */
var longestStrChain = function(words) {
    words.sort((left, right) => left.length - right.length);
    let dp = {};
    let max = 0;
    for (let word of words) {
        let len = 0;
        for (let i = 0; i < word.length; ++i) {
            let predecessor = word.slice(0, i) + word.slice(i + 1);
            len = Math.max(len, (dp[predecessor] || 0) + 1);
        }
        dp[word] = Math.max(max, len);
    }
    return Math.max(...Object.values(dp));
};
